<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Extending summarized experiments" href="extend_se.html" /><link rel="prev" title="SummarizedExperiment" href="index.html" />

    <!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>Represent genomic experimental data - SummarizedExperiment 0.5.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">SummarizedExperiment 0.5.4 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">SummarizedExperiment 0.5.4 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="extend_se.html">Extending Classes</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api/modules.html">Module Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="api/summarizedexperiment.html">summarizedexperiment package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/tutorial.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section class="tex2jax_ignore mathjax_ignore" id="represent-genomic-experimental-data">
<h1>Represent genomic experimental data<a class="headerlink" href="#represent-genomic-experimental-data" title="Link to this heading">¶</a></h1>
<p>This package provides classes to represent genomic experiments, commonly generated by sequencing experiments. It provides similar interfaces and functionality as the <a class="reference external" href="https://github.com/Bioconductor/SummarizedExperiment">Bioconductor equivalent</a>. The package currently provides both <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> &amp; <code class="docutils literal notranslate"><span class="pre">RangedSummarizedExperiment</span></code> representations. A key difference between these two is the <code class="docutils literal notranslate"><span class="pre">RangedSummarizedExperiment</span></code> object provides an additional slot to store genomic regions for each feature and is expected to be <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> (more <a class="reference external" href="https://github.com/BiocPy/GenomicRanges/">here</a>).</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The design of <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> class and its derivates adheres to the R/Bioconductor specification, where rows correspond to features, and columns represent samples or cells.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These classes follow a functional paradigm for accessing or setting properties, with further details discussed in <a class="reference external" href="https://biocpy.github.io/tutorial/chapters/philosophy.html#functional-discipline">functional paradigm</a> section.</p>
</div>
<section id="construct-a-summarizedexperiment-object">
<h2>Construct a <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> object<a class="headerlink" href="#construct-a-summarizedexperiment-object" title="Link to this heading">¶</a></h2>
<p>A <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> contains three key attributes,</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">assays</span></code>: A dictionary of matrices with assay names as keys, e.g. counts, logcounts etc.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">row_data</span></code>: Feature information e.g. genes, transcripts, exons, etc.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">column_data</span></code>: Sample information about the columns of the matrices.</p></li>
</ul>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Both <code class="docutils literal notranslate"><span class="pre">row_data</span></code> and <code class="docutils literal notranslate"><span class="pre">column_data</span></code> are expected to be <a class="reference external" href="https://github.com/BiocPy/BiocFrame">BiocFrame</a> objects and will be coerced to a <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> for consistent downstream operations.</p>
</div>
<p>In addition, these classes can optionally accept <code class="docutils literal notranslate"><span class="pre">row_names</span></code> and <code class="docutils literal notranslate"><span class="pre">column_names</span></code>. Since <code class="docutils literal notranslate"><span class="pre">row_data</span></code> and <code class="docutils literal notranslate"><span class="pre">column_data</span></code> may also contain names, the following rules are used in the implementation:</p>
<ul class="simple">
<li><p>On <strong>construction</strong>, if <code class="docutils literal notranslate"><span class="pre">row_names</span></code> or <code class="docutils literal notranslate"><span class="pre">column_names</span></code> are not provided, these are automatically inferred from <code class="docutils literal notranslate"><span class="pre">row_data</span></code> and <code class="docutils literal notranslate"><span class="pre">column_data</span></code> objects.</p></li>
<li><p>On <strong>accessors</strong> of these objects, the <code class="docutils literal notranslate"><span class="pre">row_names</span></code> in <code class="docutils literal notranslate"><span class="pre">row_data</span></code> and <code class="docutils literal notranslate"><span class="pre">column_data</span></code> are replaced by the equivalents from the SE level.</p></li>
<li><p>On <strong>setters</strong> for these attributes, especially with the functional style (<code class="docutils literal notranslate"><span class="pre">set_row_data</span></code> and <code class="docutils literal notranslate"><span class="pre">set_column_data</span></code> methods), additional options are available to replace the names in the SE object.</p></li>
</ul>
<div class="admonition caution">
<p class="admonition-title">Caution</p>
<p>These rules help avoid unexpected mdifications in names, when either <code class="docutils literal notranslate"><span class="pre">row_data</span></code> or <code class="docutils literal notranslate"><span class="pre">column_data</span></code> objects are modified.</p>
</div>
<p>To construct a <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>, we’ll first generate a matrix of read counts, representing the read counts from a series of RNA-seq experiments. Following that, we’ll create a <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> object to denote feature information and a table for column annotations. This table may include the names for the columns and any other values we wish to represent.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">random</span><span class="w"> </span><span class="kn">import</span> <span class="n">random</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">biocframe</span><span class="w"> </span><span class="kn">import</span> <span class="n">BiocFrame</span>

<span class="n">nrows</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">ncols</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">)</span>
<span class="n">row_data</span> <span class="o">=</span> <span class="n">BiocFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="o">*</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;starts&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span>
        <span class="s2">&quot;ends&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">110</span><span class="p">,</span> <span class="mi">310</span><span class="p">),</span>
        <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">),</span>
        <span class="s2">&quot;GC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">random</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)]</span> <span class="o">*</span> <span class="mi">20</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">)</span>

<span class="n">col_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;treatment&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ChIP&quot;</span><span class="p">,</span> <span class="s2">&quot;Input&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The inputs <code class="docutils literal notranslate"><span class="pre">row_data</span></code> and <code class="docutils literal notranslate"><span class="pre">column_data</span></code> are expected to be <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> objects and will be coerced to a <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> if a pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> is supplied.</p>
</div>
<section id="summarizedexperiment">
<h3><code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code><a class="headerlink" href="#summarizedexperiment" title="Link to this heading">¶</a></h3>
<p>A <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> is a base class to represent any genomic experiment. This class expects features (<code class="docutils literal notranslate"><span class="pre">row_data</span></code>) to be either a pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> or any variant of <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">summarizedexperiment</span><span class="w"> </span><span class="kn">import</span> <span class="n">SummarizedExperiment</span>

<span class="n">se</span> <span class="o">=</span> <span class="n">SummarizedExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">counts</span><span class="p">},</span> <span class="n">row_data</span><span class="o">=</span><span class="n">row_data</span><span class="p">,</span> <span class="n">column_data</span><span class="o">=</span><span class="n">col_data</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">se</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (200, 6)
assays(1): [&#39;counts&#39;]
row_data columns(6): [&#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;]
row_names(0):  
column_data columns(1): [&#39;treatment&#39;]
column_names(6): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
<section id="rangedsummarizedexperiment">
<h3><code class="docutils literal notranslate"><span class="pre">RangedSummarizedExperiment</span></code><a class="headerlink" href="#rangedsummarizedexperiment" title="Link to this heading">¶</a></h3>
<p>Similarly, we can use the same information to construct a <code class="docutils literal notranslate"><span class="pre">RangeSummarizedExperiment</span></code>. We convert feature information into a <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object and provide this as <code class="docutils literal notranslate"><span class="pre">row_ranges</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">genomicranges</span><span class="w"> </span><span class="kn">import</span> <span class="n">GenomicRanges</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">summarizedexperiment</span><span class="w"> </span><span class="kn">import</span> <span class="n">RangedSummarizedExperiment</span>

<span class="n">gr</span> <span class="o">=</span> <span class="n">GenomicRanges</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span><span class="n">row_data</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">())</span>

<span class="n">rse</span> <span class="o">=</span> <span class="n">RangedSummarizedExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">counts</span><span class="p">},</span> <span class="n">row_data</span><span class="o">=</span><span class="n">row_data</span><span class="p">,</span> <span class="n">row_ranges</span><span class="o">=</span><span class="n">gr</span><span class="p">,</span> <span class="n">column_data</span><span class="o">=</span><span class="n">col_data</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rse</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: RangedSummarizedExperiment
dimensions: (200, 6)
assays(1): [&#39;counts&#39;]
row_data columns(6): [&#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;]
row_names(0):  
column_data columns(1): [&#39;treatment&#39;]
column_names(6): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="delayed-or-file-backed-arrays">
<h2>Delayed or file-backed arrays<a class="headerlink" href="#delayed-or-file-backed-arrays" title="Link to this heading">¶</a></h2>
<p>The general idea is that <a class="reference external" href="https://github.com/biocpy/delayedarray">DelayedArray</a>’s are a drop-in replacement for NumPy arrays, at least for <a class="reference external" href="https://github.com/BiocPy">BiocPy</a> applications.</p>
<p>For example, we can use the <code class="docutils literal notranslate"><span class="pre">DelayedArray</span></code> inside a <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">delayedarray</span>

<span class="c1"># create a delayed array, can also be file-backed</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">d</span> <span class="o">=</span> <span class="n">delayedarray</span><span class="o">.</span><span class="n">wrap</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="c1"># operate over delayed arrays</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="n">d</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">100</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
<span class="n">total</span> <span class="o">=</span> <span class="n">filtered</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">normalized</span> <span class="o">=</span> <span class="n">filtered</span> <span class="o">/</span> <span class="n">total</span>
<span class="n">transformed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">normalized</span><span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">summarizedexperiment</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">SE</span>
<span class="n">se_delayed</span> <span class="o">=</span> <span class="n">SE</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">({</span> <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">filtered</span><span class="p">,</span> <span class="s2">&quot;lognorm&quot;</span><span class="p">:</span> <span class="n">transformed</span> <span class="p">})</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se_delayed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (50, 7)
assays(2): [&#39;counts&#39;, &#39;lognorm&#39;]
row_data columns(0): []
row_names(0):  
column_data columns(0): []
column_names(0):  
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
<section id="interop-with-anndata">
<h2>Interop with <code class="docutils literal notranslate"><span class="pre">anndata</span></code><a class="headerlink" href="#interop-with-anndata" title="Link to this heading">¶</a></h2>
<p>Converting a <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> to an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> representation is straightforward:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">se</span><span class="o">.</span><span class="n">to_anndata</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 6 × 200
    obs: &#39;treatment&#39;, &#39;rownames&#39;
    var: &#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SummarizedExperiment/SummarizedExperiment/.tox/docs/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
</div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>To convert an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object to a BiocPy representation, utilize the <code class="docutils literal notranslate"><span class="pre">from_anndata</span></code> method in the <a class="reference external" href="https://github.com/BiocPy/SingleCellExperiment">SingleCellExperiment</a> class. This minimizes the loss of information when converting between these two representations.</p>
</div>
</section>
<section id="getters-setters">
<h2>Getters/Setters<a class="headerlink" href="#getters-setters" title="Link to this heading">¶</a></h2>
<p>Getters are available to access various attributes using either the property notation or functional style.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># access assay names</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;assay names (as property): &quot;</span><span class="p">,</span> <span class="n">se</span><span class="o">.</span><span class="n">assay_names</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;assay names (functional style): &quot;</span><span class="p">,</span> <span class="n">se</span><span class="o">.</span><span class="n">get_assay_names</span><span class="p">())</span>

<span class="c1"># access row data</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se</span><span class="o">.</span><span class="n">row_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>assay names (as property):  [&#39;counts&#39;]
assay names (functional style):  [&#39;counts&#39;]
BiocFrame with 200 rows and 6 columns
      seqnames  starts    ends strand   score                  GC
        &lt;list&gt; &lt;range&gt; &lt;range&gt; &lt;list&gt; &lt;range&gt;              &lt;list&gt;
  [0]     chr1     100     110      -       0  0.3198107010518143
  [1]     chr2     101     111      +       1  0.8483567852533034
  [2]     chr2     102     112      +       2  0.9219032677995332
           ...     ...     ...    ...     ...                 ...
[197]     chr3     297     307      +     197  0.8942271983924933
[198]     chr3     298     308      -     198 0.44201556588285573
[199]     chr3     299     309      -     199  0.7404709415323198
</pre></div>
</div>
</div>
</div>
<section id="access-an-assay">
<h3>Access an assay<a class="headerlink" href="#access-an-assay" title="Link to this heading">¶</a></h3>
<p>One can access an assay by index or name:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">se</span><span class="o">.</span><span class="n">assay</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="c1"># same as se.assay(&quot;counts&quot;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[0.51429341, 0.54184023, 0.16596558, 0.78264256, 0.62623082,
        0.61035478],
       [0.79468695, 0.36262979, 0.45120988, 0.59818002, 0.89219842,
        0.08470789],
       [0.19467352, 0.96918153, 0.30564327, 0.83974825, 0.02334665,
        0.10654715],
       ...,
       [0.90762617, 0.35557496, 0.7674455 , 0.59779049, 0.04495346,
        0.61621557],
       [0.55149653, 0.86267911, 0.88776225, 0.6827858 , 0.65211024,
        0.86912263],
       [0.17521157, 0.68291002, 0.6732641 , 0.89520667, 0.83360793,
        0.1807409 ]], shape=(200, 6))
</pre></div>
</div>
</div>
</div>
</section>
<section id="setters">
<h3>Setters<a class="headerlink" href="#setters" title="Link to this heading">¶</a></h3>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>All property-based setters are <code class="docutils literal notranslate"><span class="pre">in_place</span></code> operations, with further details discussed in <a class="reference external" href="https://biocpy.github.io/tutorial/chapters/philosophy.html#functional-discipline">functional paradigm</a> section.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">modified_column_data</span> <span class="o">=</span> <span class="n">se</span><span class="o">.</span><span class="n">column_data</span><span class="o">.</span><span class="n">set_column</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">16</span><span class="p">))</span>
<span class="n">modified_se</span> <span class="o">=</span> <span class="n">se</span><span class="o">.</span><span class="n">set_column_data</span><span class="p">(</span><span class="n">modified_column_data</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">modified_se</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (200, 6)
assays(1): [&#39;counts&#39;]
row_data columns(6): [&#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;]
row_names(0):  
column_data columns(2): [&#39;treatment&#39;, &#39;score&#39;]
column_names(6): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
<p>Now, lets check the <code class="docutils literal notranslate"><span class="pre">column_data</span></code> on the original object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">se</span><span class="o">.</span><span class="n">column_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>BiocFrame with 6 rows and 1 column
  treatment
     &lt;list&gt;
0      ChIP
1     Input
2      ChIP
3     Input
4      ChIP
5     Input
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="subset-experiments">
<h2>Subset experiments<a class="headerlink" href="#subset-experiments" title="Link to this heading">¶</a></h2>
<p>You can subset experimental data by using the subset (<code class="docutils literal notranslate"><span class="pre">[]</span></code>) operator. This operation accepts different slice input types, such as a boolean vector, a <code class="docutils literal notranslate"><span class="pre">slice</span></code> object, a list of indices, or names (if available) to subset.</p>
<p>In our previous example, we didn’t include row or column names. Let’s create another <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> object that includes names.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">row_data</span> <span class="o">=</span> <span class="n">BiocFrame</span><span class="p">({</span>
    <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr_5&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_3&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_2&quot;</span><span class="p">],</span>
    <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">300</span><span class="p">],</span>
    <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">110</span><span class="p">,</span> <span class="mi">210</span><span class="p">,</span> <span class="mi">310</span><span class="p">],</span>
<span class="p">})</span>

<span class="n">col_data</span> <span class="o">=</span> <span class="n">BiocFrame</span><span class="p">({</span>
    <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SAM_1&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_3&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_3&quot;</span><span class="p">],</span>
    <span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">row_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_1&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_2&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_3&quot;</span><span class="p">],</span>
<span class="p">)</span>

<span class="n">se_with_names</span> <span class="o">=</span> <span class="n">SummarizedExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="s2">&quot;lognorm&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">lognormal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">row_data</span><span class="p">,</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">col_data</span><span class="p">,</span>
    <span class="n">row_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HER2&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
    <span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_1&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_2&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_3&quot;</span><span class="p">],</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">se_with_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (3, 3)
assays(2): [&#39;counts&#39;, &#39;lognorm&#39;]
row_data columns(3): [&#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;]
row_names(3): [&#39;HER2&#39;, &#39;BRCA1&#39;, &#39;TPFK&#39;]
column_data columns(2): [&#39;sample&#39;, &#39;disease&#39;]
column_names(3): [&#39;cell_1&#39;, &#39;cell_2&#39;, &#39;cell_3&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
<section id="subset-by-index-position">
<h3>Subset by index position<a class="headerlink" href="#subset-by-index-position" title="Link to this heading">¶</a></h3>
<p>A straightforward slice operation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subset_se</span> <span class="o">=</span> <span class="n">se_with_names</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">subset_se</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (3, 3)
assays(2): [&#39;counts&#39;, &#39;lognorm&#39;]
row_data columns(3): [&#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;]
row_names(3): [&#39;HER2&#39;, &#39;BRCA1&#39;, &#39;TPFK&#39;]
column_data columns(2): [&#39;sample&#39;, &#39;disease&#39;]
column_names(3): [&#39;cell_1&#39;, &#39;cell_2&#39;, &#39;cell_3&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
<section id="subset-by-row-names-or-column-names">
<h3>Subset by row names or column names<a class="headerlink" href="#subset-by-row-names-or-column-names" title="Link to this heading">¶</a></h3>
<p>Either one or both of the slices can contain names. These names are mapped to <code class="docutils literal notranslate"><span class="pre">row_names</span></code> and <code class="docutils literal notranslate"><span class="pre">column_names</span></code> of the <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subset_se</span> <span class="o">=</span> <span class="n">se_with_names</span><span class="p">[:</span><span class="mi">2</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;cell_1&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_3&quot;</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">subset_se</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (2, 2)
assays(2): [&#39;counts&#39;, &#39;lognorm&#39;]
row_data columns(3): [&#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;]
row_names(2): [&#39;HER2&#39;, &#39;BRCA1&#39;]
column_data columns(2): [&#39;sample&#39;, &#39;disease&#39;]
column_names(2): [&#39;cell_1&#39;, &#39;cell_3&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
<p>An <code class="docutils literal notranslate"><span class="pre">Exception</span></code> is raised if a names does not exist.</p>
</section>
<section id="subset-by-boolean-vector">
<h3>Subset by boolean vector<a class="headerlink" href="#subset-by-boolean-vector" title="Link to this heading">¶</a></h3>
<p>Similarly, you can also slice by a boolean array.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Note that the boolean vectors should contain the same number of features for the row slice and the same number of samples for the column slice.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subset_se_with_bools</span> <span class="o">=</span> <span class="n">se_with_names</span><span class="p">[[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">subset_se_with_bools</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (2, 2)
assays(2): [&#39;counts&#39;, &#39;lognorm&#39;]
row_data columns(3): [&#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;]
row_names(2): [&#39;HER2&#39;, &#39;BRCA1&#39;]
column_data columns(2): [&#39;sample&#39;, &#39;disease&#39;]
column_names(2): [&#39;cell_1&#39;, &#39;cell_3&#39;]
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
<section id="subset-by-empty-list">
<h3>Subset by empty list<a class="headerlink" href="#subset-by-empty-list" title="Link to this heading">¶</a></h3>
<p>This is a feature not a bug ;), you can specify an empty list to completely remove all rows or samples.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>An empty array (<code class="docutils literal notranslate"><span class="pre">[]</span></code>) is not the same as an empty slice (<code class="docutils literal notranslate"><span class="pre">:</span></code>). This helps us avoid unintented operations.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subset</span> <span class="o">=</span> <span class="n">se_with_names</span><span class="p">[:</span><span class="mi">2</span><span class="p">,</span> <span class="p">[]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SummarizedExperiment
dimensions: (2, 0)
assays(2): [&#39;counts&#39;, &#39;lognorm&#39;]
row_data columns(3): [&#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;]
row_names(2): [&#39;HER2&#39;, &#39;BRCA1&#39;]
column_data columns(2): [&#39;sample&#39;, &#39;disease&#39;]
column_names(0): []
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
<section id="range-based-operations">
<h3>Range-based operations<a class="headerlink" href="#range-based-operations" title="Link to this heading">¶</a></h3>
<p>Additionally, since <code class="docutils literal notranslate"><span class="pre">RangeSummarizedExperiment</span></code> contains <code class="docutils literal notranslate"><span class="pre">row_ranges</span></code>, this allows us to perform a number of range-based operations that are possible on a <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
<p>For example, to subset <code class="docutils literal notranslate"><span class="pre">RangeSummarizedExperiment</span></code> with a <strong>query</strong> set of regions:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">iranges</span><span class="w"> </span><span class="kn">import</span> <span class="n">IRanges</span>
<span class="n">query</span> <span class="o">=</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="n">seqnames</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr2&quot;</span><span class="p">],</span> <span class="n">ranges</span><span class="o">=</span><span class="n">IRanges</span><span class="p">([</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">6</span><span class="p">]),</span> <span class="n">strand</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">])</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">rse</span><span class="o">.</span><span class="n">subset_by_overlaps</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">TypeError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">15</span><span class="p">],</span> <span class="n">line</span> <span class="mi">4</span>
<span class="g g-Whitespace">      </span><span class="mi">1</span> <span class="kn">from</span><span class="w"> </span><span class="nn">iranges</span><span class="w"> </span><span class="kn">import</span> <span class="n">IRanges</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="n">query</span> <span class="o">=</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="n">seqnames</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr2&quot;</span><span class="p">],</span> <span class="n">ranges</span><span class="o">=</span><span class="n">IRanges</span><span class="p">([</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">6</span><span class="p">]),</span> <span class="n">strand</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">])</span>
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="n">result</span> <span class="o">=</span> <span class="n">rse</span><span class="o">.</span><span class="n">subset_by_overlaps</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span> <span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="nn">File ~/work/SummarizedExperiment/SummarizedExperiment/.tox/docs/lib/python3.11/site-packages/summarizedexperiment/RangedSummarizedExperiment.py:879,</span> in <span class="ni">RangedSummarizedExperiment.subset_by_overlaps</span><span class="nt">(self, query, query_type, max_gap, min_overlap, ignore_strand)</span>
<span class="g g-Whitespace">    </span><span class="mi">875</span> <span class="n">_check_gr_or_rse</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">877</span> <span class="n">qranges</span> <span class="o">=</span> <span class="n">_access_granges</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">879</span> <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">row_ranges</span><span class="o">.</span><span class="n">find_overlaps</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">880</span>     <span class="n">query</span><span class="o">=</span><span class="n">qranges</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">881</span>     <span class="n">query_type</span><span class="o">=</span><span class="n">query_type</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">882</span>     <span class="n">max_gap</span><span class="o">=</span><span class="n">max_gap</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">883</span>     <span class="n">min_overlap</span><span class="o">=</span><span class="n">min_overlap</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">884</span>     <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">885</span> <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">887</span> <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">result</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;self_hits&quot;</span><span class="p">),</span> <span class="p">:]</span>

<span class="nn">File ~/work/SummarizedExperiment/SummarizedExperiment/.tox/docs/lib/python3.11/site-packages/genomicranges/GenomicRanges.py:2233,</span> in <span class="ni">GenomicRanges.find_overlaps</span><span class="nt">(self, query, query_type, select, max_gap, min_overlap, ignore_strand)</span>
<span class="g g-Whitespace">   </span><span class="mi">2222</span>         <span class="n">_query_subset</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>
<span class="g g-Whitespace">   </span><span class="mi">2224</span>         <span class="n">res_idx</span> <span class="o">=</span> <span class="n">_sub_subset</span><span class="o">.</span><span class="n">_ranges</span><span class="o">.</span><span class="n">find_overlaps</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">2225</span>             <span class="n">query</span><span class="o">=</span><span class="n">_query_subset</span><span class="o">.</span><span class="n">_ranges</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2226</span>             <span class="n">query_type</span><span class="o">=</span><span class="n">query_type</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>   <span class="mi">2230</span>             <span class="n">delete_index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2231</span>         <span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">2233</span>         <span class="n">all_qhits</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">2234</span>             <span class="p">(</span><span class="n">all_qhits</span><span class="p">,</span> <span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">res_idx</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;query_hits&quot;</span><span class="p">)]),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span>
<span class="g g-Whitespace">   </span><span class="mi">2235</span>         <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">2236</span>         <span class="n">all_shits</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">2237</span>             <span class="p">(</span><span class="n">all_shits</span><span class="p">,</span> <span class="p">[</span><span class="n">_subset</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res_idx</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;self_hits&quot;</span><span class="p">)]),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span>
<span class="g g-Whitespace">   </span><span class="mi">2238</span>         <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">2240</span> <span class="n">order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">all_qhits</span><span class="p">,</span> <span class="n">stable</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="ne">TypeError</span>: Cannot cast array data from dtype(&#39;float64&#39;) to dtype(&#39;int32&#39;) according to the rule &#39;same_kind&#39;
</pre></div>
</div>
</div>
</div>
<p>Additionally, RSE supports many other interval based operations. Checkout the <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/modules.html">documentation</a> for more details.</p>
</section>
</section>
<section id="combining-experiments">
<h2>Combining experiments<a class="headerlink" href="#combining-experiments" title="Link to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> implements methods for the <code class="docutils literal notranslate"><span class="pre">combine</span></code> generics from <a class="reference external" href="https://github.com/BiocPy/biocutils"><strong>BiocUtils</strong></a>.</p>
<p>These methods enable the merging or combining of multiple <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> objects, allowing users to aggregate data from different experiments or conditions. To demonstrate, let’s create multiple <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> objects.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rowData1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr_5&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_3&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10293804</span><span class="p">,</span> <span class="mi">12098948</span><span class="p">,</span> <span class="mi">20984392</span><span class="p">],</span>
        <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">28937947</span><span class="p">,</span> <span class="mi">3872839</span><span class="p">,</span> <span class="mi">329837492</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HER2&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">colData1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SAM_1&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_3&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_3&quot;</span><span class="p">],</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_1&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_2&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_3&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">se1</span> <span class="o">=</span> <span class="n">SummarizedExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="s2">&quot;lognorm&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">lognormal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">rowData1</span><span class="p">,</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">colData1</span><span class="p">,</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq_type&quot;</span><span class="p">:</span> <span class="s2">&quot;paired&quot;</span><span class="p">},</span>
<span class="p">)</span>

<span class="n">rowData2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr_5&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_3&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10293804</span><span class="p">,</span> <span class="mi">12098948</span><span class="p">,</span> <span class="mi">20984392</span><span class="p">],</span>
        <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">28937947</span><span class="p">,</span> <span class="mi">3872839</span><span class="p">,</span> <span class="mi">329837492</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HER2&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">colData2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SAM_4&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_5&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_6&quot;</span><span class="p">],</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;False&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_4&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_5&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_6&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">se2</span> <span class="o">=</span> <span class="n">SummarizedExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="s2">&quot;lognorm&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">lognormal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">rowData2</span><span class="p">,</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">colData2</span><span class="p">,</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq_platform&quot;</span><span class="p">:</span> <span class="s2">&quot;Illumina NovaSeq 6000&quot;</span><span class="p">},</span>
<span class="p">)</span>

<span class="n">rowData3</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr_7&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_Y&quot;</span><span class="p">],</span>
        <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1084390</span><span class="p">,</span> <span class="mi">1874937</span><span class="p">,</span> <span class="mi">243879798</span><span class="p">],</span>
        <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">243895239</span><span class="p">,</span> <span class="mi">358908298</span><span class="p">,</span> <span class="mi">390820395</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;MYC&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA2&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">colData3</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SAM_7&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_8&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_9&quot;</span><span class="p">],</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;False&quot;</span><span class="p">,</span> <span class="s2">&quot;False&quot;</span><span class="p">],</span>
        <span class="s2">&quot;doublet_score&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">.15</span><span class="p">,</span> <span class="mf">.62</span><span class="p">,</span> <span class="mf">.18</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_7&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_8&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_9&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">se3</span> <span class="o">=</span> <span class="n">SummarizedExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="s2">&quot;lognorm&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">lognormal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">rowData3</span><span class="p">,</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">colData3</span><span class="p">,</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq_platform&quot;</span><span class="p">:</span> <span class="s2">&quot;Illumina NovaSeq 6000&quot;</span><span class="p">},</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">se1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The <code class="docutils literal notranslate"><span class="pre">combine_rows</span></code> or <code class="docutils literal notranslate"><span class="pre">combine_columns</span></code> operations, expect all experiments to contain the same assay names.</p>
</div>
<p>To combine experiments by row:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">biocutils</span><span class="w"> </span><span class="kn">import</span> <span class="n">relaxed_combine_columns</span><span class="p">,</span> <span class="n">combine_columns</span><span class="p">,</span> <span class="n">combine_rows</span><span class="p">,</span> <span class="n">relaxed_combine_rows</span>
<span class="n">se_combined</span> <span class="o">=</span> <span class="n">combine_rows</span><span class="p">(</span><span class="n">se2</span><span class="p">,</span> <span class="n">se1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se_combined</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Similarly to combine by column:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">se_combined</span> <span class="o">=</span> <span class="n">combine_columns</span><span class="p">(</span><span class="n">se2</span><span class="p">,</span> <span class="n">se1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se_combined</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>You can use <code class="docutils literal notranslate"><span class="pre">relaxed_combine_columns</span></code> or <code class="docutils literal notranslate"><span class="pre">relaxed_combined_rows</span></code> when there’s mismatch in the number of features or samples. Missing rows or columns in any object are filled in with appropriate placeholder values before combining, e.g. missing assay’s are replaced with a masked numpy array.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># se3 contains an additional assay not present in se1</span>
<span class="n">se_relaxed_combine</span> <span class="o">=</span> <span class="n">relaxed_combine_columns</span><span class="p">(</span><span class="n">se3</span><span class="p">,</span> <span class="n">se1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">se_relaxed_combine</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="empty-experiments">
<h3>Empty experiments<a class="headerlink" href="#empty-experiments" title="Link to this heading">¶</a></h3>
<p>Both these classes can also contain no experimental data, and they tend to be useful when integrated into more extensive data structures but do not contain any data themselves.</p>
<p>To create an empty <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">empty_se</span> <span class="o">=</span> <span class="n">SummarizedExperiment</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">empty_se</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Similarly an empty <code class="docutils literal notranslate"><span class="pre">RangeSummarizedExperiment</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">empty_rse</span> <span class="o">=</span> <span class="n">RangedSummarizedExperiment</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">empty_rse</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="extend_se.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Extending summarized experiments</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, jkanche
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Represent genomic experimental data</a><ul>
<li><a class="reference internal" href="#construct-a-summarizedexperiment-object">Construct a <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> object</a><ul>
<li><a class="reference internal" href="#summarizedexperiment"><code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></li>
<li><a class="reference internal" href="#rangedsummarizedexperiment"><code class="docutils literal notranslate"><span class="pre">RangedSummarizedExperiment</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#delayed-or-file-backed-arrays">Delayed or file-backed arrays</a></li>
<li><a class="reference internal" href="#interop-with-anndata">Interop with <code class="docutils literal notranslate"><span class="pre">anndata</span></code></a></li>
<li><a class="reference internal" href="#getters-setters">Getters/Setters</a><ul>
<li><a class="reference internal" href="#access-an-assay">Access an assay</a></li>
<li><a class="reference internal" href="#setters">Setters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#subset-experiments">Subset experiments</a><ul>
<li><a class="reference internal" href="#subset-by-index-position">Subset by index position</a></li>
<li><a class="reference internal" href="#subset-by-row-names-or-column-names">Subset by row names or column names</a></li>
<li><a class="reference internal" href="#subset-by-boolean-vector">Subset by boolean vector</a></li>
<li><a class="reference internal" href="#subset-by-empty-list">Subset by empty list</a></li>
<li><a class="reference internal" href="#range-based-operations">Range-based operations</a></li>
</ul>
</li>
<li><a class="reference internal" href="#combining-experiments">Combining experiments</a><ul>
<li><a class="reference internal" href="#empty-experiments">Empty experiments</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=e1a677a5"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>